compiler/damlc/tests/src/DA/Test/ShakeIdeClient.hs

```diff
@@ -409,6 +409,107 @@ dlintSmokeTests mbScenarioService = Tasty.testGroup "Dlint smoke tests"
             setFilesOfInterest [foo]
             expectNoErrors
             expectDiagnostic DsInfo (foo, 2, 11) "Suggestion: Use ::"
+    ,  testCase' "Use guards" $ do
+            foo <- makeFile "Foo.daml" $ T.unlines
+                [ "daml 1.2"
+                , "module Foo where"
+                , "truth i = if i == 1 then Some True else if i == 2 then Some False else None"
+                ]
+            setFilesOfInterest [foo]
+            expectNoErrors
+            expectDiagnostic DsInfo (foo, 2, 0) "Suggestion: Use guards"
+    ,  testCase' "Redundant guard" $ do
+            foo <- makeFile "Foo.daml" $ T.unlines
+                [ "daml 1.2"
+                , "module Foo where"
+                , "foo i | otherwise = True"
+                ]
+            setFilesOfInterest [foo]
+            expectNoErrors
+            expectDiagnostic DsInfo (foo, 2, 0) "Suggestion: Redundant guard"
+    ,  testCase' "Redundant where" $ do
+            foo <- makeFile "Foo.daml" $ T.unlines
+                [ "daml 1.2"
+                , "module Foo where"
+                , "foo i = i where"
+                ]
+            setFilesOfInterest [foo]
+            expectNoErrors
+            expectDiagnostic DsInfo (foo, 2, 0) "Suggestion: Redundant where"
+    ,  testCase' "Use otherwise" $ do
+            foo <- makeFile "Foo.daml" $ T.unlines
+                [ "daml 1.2"
+                , "module Foo where"
+                , "foo i | i == 1 = True | True = False"
+                ]
+            setFilesOfInterest [foo]
+            expectNoErrors
+            expectDiagnostic DsInfo (foo, 2, 0) "Suggestion: Use otherwise"
+    ,  testCase' "Use record patterns" $ do
+            foo <- makeFile "Foo.daml" $ T.unlines
+                [ "daml 1.2"
+                , "module Foo where"
+                , "data Foo = Foo {a : Int, b : Int, c : Int, d : Int}"
+                , "foo (Foo _ _ _ _) = True"
```

Comment:
Is there a `with`-syntax version of record patterns? Given that @bame-da seems to be fairly keen on using `with` syntax in all ouf our docs (and I agree that being consistent is better), I’m a bit hesistant to suggest braces by default.

Reply:
Do you mean that the suggestion (in DAML), should read `Suggest : Foo with {...}`?

Reply:
I don’t know what the `with` syntax of `foo Foo {} = …` is, I guess ` foo Foo with {} = `? Doesn’t seem like an improvement to me but I’m the wrong person to comment on the beauty of with-syntax.

Reply:
if you need more i can refer you to a helpful blog post (disclaimer : i'm the author!) 🤣

Reply:
leaving :trollface: mode for a moment, suggest letting it slide since it's only a shake ide test. the question above though - perhaps our suggestion should read `Suggest : Foo with {...}`?

Reply:
it's a minor enhancement but one more thing to maintain on the fork (not a bad thing necessarily - just undecided about that - generally try to keep it restricted to that which is mandatory e.g. different parse trees)

Reply:
It’s not just a Shake IDE test, right? This is a rule that DLint will suggest by default.

Reply:
`Foo with {..}` looks like a record wild card which is different from `Foo {}`.

Reply:
Sorry, I didn't mean that. Should have wrote,
```
Foo {}
```
I'm confused though. Can we back up? What is it you are advocating should happen here? Is it that the message should read `Suggest : Foo with {}`?

Reply:
I’m not sure I’m advocating for that but it does at least seem worth considering given that we try to guide users towards `with`-syntax.

Reply:
Ok I got you. Yes, I can get onboard with that and it's simple enough to implement. Let's do it.